src/mango/src/mango_idx_nouveau.erl

```diff
@@ -0,0 +1,459 @@
+    end;
+indexable_fields(Fields, {op_or, Args}) when is_list(Args) ->
+    lists:foldl(
+        fun(Arg, Fields0) -> indexable_fields(Fields0, Arg) end,
+        Fields,
+        Args
+    );
+indexable_fields(Fields, {op_not, {ExistsQuery, Arg}}) when is_tuple(Arg) ->
+    Fields0 = indexable_fields(Fields, ExistsQuery),
+    indexable_fields(Fields0, Arg);
+% forces "$exists" : false to use _all_docs
+indexable_fields(_, {op_not, {_, false}}) ->
+    [];
+indexable_fields(Fields, {op_insert, Arg}) when is_binary(Arg) ->
+    Fields;
+%% fieldname.[]:length is not a user defined field.
+indexable_fields(Fields, {op_field, {[_, <<":length">>], _}}) ->
+    Fields;
+indexable_fields(Fields, {op_field, {Name, _}}) ->
+    [iolist_to_binary(Name) | Fields];
+%% In this particular case, the lucene index is doing a field_exists query
+%% so it is looking at all sorts of combinations of field:* and field.*
+%% We don't add the field because we cannot pre-determine what field will exist.
+%% Hence we just return Fields and make it less restrictive.
+indexable_fields(Fields, {op_fieldname, {_, _}}) ->
+    Fields;
+%% Similar idea to op_fieldname but with fieldname:null
+indexable_fields(Fields, {op_null, {_, _}}) ->
+    Fields;
+indexable_fields(Fields, {op_default, _}) ->
+    [<<"$default">> | Fields].
+
+maybe_reject_index_all_req({Def}, Db) ->
+    DbName = couch_db:name(Db),
+    #user_ctx{name = User} = couch_db:get_user_ctx(Db),
+    Fields = couch_util:get_value(fields, Def),
+    case {Fields, forbid_index_all()} of
+        {all_fields, "true"} ->
+            ?MANGO_ERROR(index_all_disabled);
+        {all_fields, "warn"} ->
+            couch_log:warning(
+                "User ~p is indexing all fields in db ~p",
+                [User, DbName]
+            );
+        _ ->
+            ok
+    end.
+
+forbid_index_all() ->
+    config:get("mango", "index_all_disabled", "false").
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+
+setup_all() ->
+    Ctx = test_util:start_couch(),
+    meck:expect(
+        couch_log,
+        warning,
+        2,
+        fun(_, _) ->
+            throw({test_error, logged_warning})
+        end
+    ),
+    Ctx.
+
+teardown_all(Ctx) ->
+    meck:unload(),
+    test_util:stop_couch(Ctx).
+
+setup() ->
+    %default index all def that generates {fields, all_fields}
+    Index = #idx{def = {[]}},
+    DbName = <<"testdb">>,
+    UserCtx = #user_ctx{name = <<"u1">>},
+    {ok, Db} = couch_db:clustered_db(DbName, UserCtx),
+    {Index, Db}.
+
+teardown(_) ->
+    ok.
+
+index_all_test_() ->
+    {
+        setup,
+        fun setup_all/0,
+        fun teardown_all/1,
+        {
+            foreach,
+            fun setup/0,
+            fun teardown/1,
+            [
+                fun forbid_index_all/1,
+                fun default_and_false_index_all/1,
+                fun warn_index_all/1
+            ]
+        }
+    }.
+
+forbid_index_all({Idx, Db}) ->
+    ?_test(begin
```

Comment:
Avoid using `?_test(begin ... end)` construct, use the `?TDEF_FE` macro instead. See https://github.com/apache/couchdb/blob/main/src/chttpd/test/eunit/chttpd_dbs_info_test.erl#L70-L83 as an example or search for `?TDEF_FE`.

Reply:
hm, these tests were copied from the idx_text, they're probably not useful in context.